# Exercise - Data Exploration

Load the data
- 이 섹션에서는 판다 라이브러리를 사용
- 가장 먼저 해야 할 일은 CSV 데이터를 로드하고 읽을 수 있는지 확인하는 것
- Pandas 라이브러리에는 빠른 CSV 로더가 있으며 그 결과는 데이터 프레임에 저장됨
- 로드하는 CSV에는 50만 개가 넘는 행이 있지만 열은 17개에 불과
- Pandas는 모든 행에 대해 작업을 수행하는 기능을 포함하여 데이터 프레임과 상호 작용할 수 있는 다양한 방법을 제공

사용할 데이터 파일을 로드하는 것부터 시작

11111231

In [1]:
# Load the hotel reviews from CSV
import pandas as pd
import time

# importing time so the start and end time can be used to calculate file loading time
print("Loading data file now, this could take a while depending on file size")
start = time.time()
df = pd.read_csv('./wine_review.csv')
end = time.time()
print("Loading took " + str(round(end - start, 2)) + " seconds")

Loading data file now, this could take a while depending on file size
Loading took 0.12 seconds


# Explore the data

- 데이터는 깨끗한 상태이므로 작업할 준비가 되어 있으며, 영어 문자만 기대하는 알고리즘을 방해하는 다른 언어의 문자가 없음
- 데이터가 로드되면 코드로 데이터를 탐색할 수 있는지 확인
- 부정적_리뷰 및 긍정적_리뷰 열은 NLP 알고리즘이 처리할 수 있는 텍스트로 채워져 있음
- NLP와 감성을 살펴보기 전에 아래 코드를 사용해서 데이터 세트에 주어진 값이 팬더로 계산한 값과 일치하는지 확인하는 것이 필요

In [4]:
# 데이터 구조 확인
df.info()

# 데이터 샘플 확인
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2890 entries, 0 to 2889
Data columns (total 32 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    2890 non-null   object 
 1   asins                 870 non-null    object 
 2   brand                 2825 non-null   object 
 3   categories            2890 non-null   object 
 4   dateAdded             2890 non-null   object 
 5   dateUpdated           2890 non-null   object 
 6   descriptions          2738 non-null   object 
 7   dimension             1838 non-null   object 
 8   ean                   2166 non-null   object 
 9   flavors               151 non-null    object 
 10  keys                  2890 non-null   object 
 11  manufacturer          849 non-null    object 
 12  manufacturerNumber    2457 non-null   object 
 13  name                  2890 non-null   object 
 14  reviews.date          2552 non-null   object 
 15  reviews.dateAdded    

id asins            brand  \
0  AV13ClKCGV-KLJ3akN68   NaN            Gallo   
1  AV13CsvW-jtxr-f38AQO   NaN  Fresh Craft Co.   
2  AV13CVI_glJLPUi8O7Po   NaN     1000 Stories   
3  AV13CVI_glJLPUi8O7Po   NaN     1000 Stories   
4  AV13CYL4-jtxr-f37_-t   NaN     Wine Cube153   

                                          categories             dateAdded  \
0  Food & Beverage,Beverages,Wine, Beer & Liquor,...  2017-07-24T23:59:11Z   
1  Food & Beverage,Beverages,Wine, Beer & Liquor,...  2017-07-24T23:59:42Z   
2  Food & Beverage,Beverages,Wine, Beer & Liquor,...  2017-07-24T23:58:05Z   
3  Food & Beverage,Beverages,Wine, Beer & Liquor,...  2017-07-24T23:58:05Z   
4  Food & Beverage,Beverages,Wine, Beer & Liquor,...  2017-07-24T23:58:18Z   

            dateUpdated                                       descriptions  \
0  2018-01-10T18:06:28Z                                                NaN   
1  2018-01-10T05:38:33Z  [{"dateSeen":["2017-12-21T05:43:00.000Z","2017...   
2  2018-01-10T05:38:31Z                                                NaN   
3  2018-01-10T05:38:31Z                                                NaN   
4  2018-01-10T18:06:29Z  [{"dateSeen":["2017-12-21T05:43:00.000Z","2017...   

                     dimension  ean flavors  ...  \
0     1.0 in x 1.0 in x 1.0 in  NaN     NaN  ...   
1  4.25 in x 4.25 in x 5.25 in  NaN     NaN  ...   
2   3.3 in x 3.3 in x 11.79 in  NaN     NaN  ...   
3   3.3 in x 3.3 in x 11.79 in  NaN     NaN  ...   
4     1.0 in x 1.0 in x 1.0 in  NaN     NaN  ...   

                                  reviews.sourceURLs  \
0  https://redsky.target.com/groot-domain-api/v1/...   
1  https://redsky.target.com/groot-domain-api/v1/...   
2  https://redsky.target.com/groot-domain-api/v1/...   
3  https://redsky.target.com/groot-domain-api/v1/...   
4  https://redsky.target.com/groot-domain-api/v1/...   

                                        reviews.text  \
0      This a fantastic white wine for any occasion!   
1   Tart, not sweet...very refreshing and delicious!   
2  I was given this wine so it was a delightful s...   
3  This is a phenomenal wine and my new favorite ...   
4  4 750ml bottles for the price of two With way ...   

                          reviews.title reviews.userCity reviews.userProvince  \
0                My Favorite White Wine              NaN                  NaN   
1                                 Yum!!              NaN                  NaN   
2                       A New Favorite!              NaN                  NaN   
3  Bold, Flavorful, Aromatic, Delicious              NaN                  NaN   
4  Yum! Plus, Environmentally Friendly!    Overland Park                  NaN   

  reviews.username sizes                                         sourceURLs  \
0              Bjh   NaN  http://redsky.target.com/v1/plp/search?kwr=y&c...   
1             Wino   NaN  http://redsky.target.com/v1/plp/search?kwr=y&c...   
2         Bama Mom   NaN  http://redsky.target.com/v1/plp/search?kwr=y&c...   
3           Av Dub   NaN  http://redsky.target.com/v1/plp/search?kwr=y&c...   
4       Chelseamay   NaN  https://redsky.target.com/groot-domain-api/v1/...   

           upc    weight  
0   4.9213E+11   1.0 lbs  
1  83120003441  2.45 lbs  
2  82896001453  3.09 lbs  
3  82896001453  3.09 lbs  
4  85200600465   1.0 lbs  

[5 rows x 32 columns]

데이터 구조 및 샘플 분석
총 32개의 컬럼과 2890개의 레코드가 있습니다.
주요 텍스트 컬럼은 reviews.text로, 이는 감성 분석에 사용될 텍스트 데이터를 포함합니다.
일부 컬럼에 결측값이 많이 포함되어 있습니다 (예: asins, dimension, flavors 등).

In [6]:
# 감성 분석에 필요한 컬럼만 선택
df = df[['reviews.text']]

# 결측값 확인
df.isnull().sum()

reviews.text    1
dtype: int64

In [8]:
# 결측값 제거
df = df.dropna(subset=['reviews.text'])

# 데이터 구조 확인
df.info()

# 데이터 샘플 확인
df.head()

<class 'pandas.core.frame.DataFrame'>
Index: 2889 entries, 0 to 2889
Data columns (total 1 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   reviews.text  2889 non-null   object
dtypes: object(1)
memory usage: 45.1+ KB


reviews.text
0      This a fantastic white wine for any occasion!
1   Tart, not sweet...very refreshing and delicious!
2  I was given this wine so it was a delightful s...
3  This is a phenomenal wine and my new favorite ...
4  4 750ml bottles for the price of two With way ...

# Dataframe operations

1. 로드한 데이터 프레임의 모양을 프린트 (모양은 행과 열의 수)
2. 검토자 국적의 빈도 수를 계산:
- 검토자_국적 열에 고유한 값이 몇 개 있으며 그 값은 무엇인가?
- 데이터 집합(인쇄 국가 및 리뷰 수)에서 가장 일반적인 리뷰어 국적은 무엇인가?
- 다음으로 가장 자주 발견되는 상위 10개 국적은 무엇이며 그 빈도 수는 얼마인가?
3. 리뷰가 가장 많이 작성된 상위 10개 국적별로 가장 자주 리뷰가 작성된 호텔은 어디인가?
4. 데이터 세트에서 호텔당 리뷰 수(호텔의 빈도 수)는 몇 개인가?
5. 데이터 집합에 각 호텔에 대한 평균 점수 열이 있지만 평균 점수(각 호텔에 대한 데이터 집합의 모든 리뷰어 점수의 평균을 구하는 것)를 계산할 수도 있음
- 계산된 평균을 포함하는 열 머리글이 Calc_Average_Score인 새 열을 데이터 프레임에 추가
6. 동일한(소수점 이하 1자리 반올림) Average_Score와 Calc_Average_Score를 가진 호텔이 있는지 확인
- Series(행)를 인수로 받아 값을 비교하고 값이 같지 않으면 메시지를 출력하는 Python 함수를 작성하고  .apply() 메서드를 사용하여 함수로 모든 행을 처리
7. 부정_검토 열 값이 "No Negative"인 행의 개수를 계산하여 출력
8. Positive_Review 열 값이 "No Positive"인 행의 개수를 계산하여 출력
9. 열 Positive_Review 값이 "No Positive", Negative_Review 값이 "No Negative"인 행의 수를 계산하여 출력

In [2]:
# Print out the shape of the loaded data frame (the shape is the number of rows and columns)
print("The shape of the data (rows, cols) is " + str(df.shape))

The shape of the data (rows, cols) is (2890, 32)


In [3]:
# Calculate the frequency count for reviewer nationalities
# How many distinct values are there for the column Reviewer_Nationality and what are they?
# What reviewer nationality is the most common in the dataset (print country and number of reviews)?
nationality_freq = df["Reviewer_Nationality"].value_counts()
print("There are " + str(nationality_freq.size) + " different nationalities")
# print first and last rows of the Series. Change to nationality_freq.to_string() to print all of the data
print(nationality_freq) 

KeyError: 'Reviewer_Nationality'

In [4]:
# What are the next top 10 most frequently found nationalities, and their frequency count?

print("The highest frequency reviewer nationality is " + str(nationality_freq.index[0]).strip() + " with " + str(nationality_freq[0]) + " reviews.")

# Notice there is a leading space on the values, strip() removes that for printing
# What is the top 10 most common nationalities and their frequencies?

print("The next 10 highest frequency reviewer nationalities are:")
print(nationality_freq[1:11].to_string())

The highest frequency reviewer nationality is United Kingdom with 245246 reviews.
The next 10 highest frequency reviewer nationalities are:
Reviewer_Nationality
 United States of America     35437
 Australia                    21686
 Ireland                      14827
 United Arab Emirates         10235
 Saudi Arabia                  8951
 Netherlands                   8772
 Switzerland                   8678
 Germany                       7941
 Canada                        7894
 France                        7296


In [5]:
# What was the most frequently reviewed hotel for the top 10 nationalities
for nat in nationality_freq[:10].index:
   # First, extract all the rows that match the criteria into a new dataframe
   nat_df = df[df["Reviewer_Nationality"] == nat]   
   # Now get the hotel freq
   freq = nat_df["Hotel_Name"].value_counts()
   print("The most reviewed hotel for " + str(nat).strip() + " was " + str(freq.index[0]) + " with " + str(freq[0]) + " reviews.") 

The most reviewed hotel for United Kingdom was Britannia International Hotel Canary Wharf with 3833 reviews.
The most reviewed hotel for United States of America was Hotel Esther a with 423 reviews.
The most reviewed hotel for Australia was Park Plaza Westminster Bridge London with 167 reviews.
The most reviewed hotel for Ireland was Copthorne Tara Hotel London Kensington with 239 reviews.
The most reviewed hotel for United Arab Emirates was Millennium Hotel London Knightsbridge with 129 reviews.
The most reviewed hotel for Saudi Arabia was The Cumberland A Guoman Hotel with 142 reviews.
The most reviewed hotel for Netherlands was Jaz Amsterdam with 97 reviews.
The most reviewed hotel for Switzerland was Hotel Da Vinci with 97 reviews.
The most reviewed hotel for Germany was Hotel Da Vinci with 86 reviews.
The most reviewed hotel for Canada was St James Court A Taj Hotel London with 61 reviews.


In [6]:
# How many reviews are there per hotel (frequency count of hotel) in the dataset?

# First create a new dataframe based on the old one, removing the uneeded columns
hotel_freq_df = df.drop(["Hotel_Address", "Additional_Number_of_Scoring", "Review_Date", "Average_Score", "Reviewer_Nationality", "Negative_Review", "Review_Total_Negative_Word_Counts", "Positive_Review", "Review_Total_Positive_Word_Counts", "Total_Number_of_Reviews_Reviewer_Has_Given", "Reviewer_Score", "Tags", "days_since_review", "lat", "lng"], axis = 1)

# Group the rows by Hotel_Name, count them and put the result in a new column Total_Reviews_Found
hotel_freq_df['Total_Reviews_Found'] = hotel_freq_df.groupby('Hotel_Name').transform('count')

# Get rid of all the duplicated rows
hotel_freq_df = hotel_freq_df.drop_duplicates(subset = ["Hotel_Name"])
display(hotel_freq_df) 


Hotel_Name  Total_Number_of_Reviews  \
0                            Hotel Arena                     1403   
405                     K K Hotel George                     1831   
971              Apex Temple Court Hotel                     2619   
2008    The Park Grand London Paddington                     4380   
3778                 Monhotel Lounge SPA                      171   
...                                  ...                      ...   
511962        Suite Hotel 900 m zur Oper                     3461   
512401                     Hotel Amadeus                      717   
512545                      The Berkeley                      232   
512645     Holiday Inn London Kensington                     5945   
515413             Atlantis Hotel Vienna                     2823   

        Total_Reviews_Found  
0                       405  
405                     566  
971                    1037  
2008                   1770  
3778                     35  
...                     ...  
511962                  439  
512401                  144  
512545                  100  
512645                 2768  
515413                  325  

[1492 rows x 3 columns]

- The counted in the dataset results do not match the value in Total_Number_of_Reviews.
- It is unclear if this value in the dataset represented the total number of reviews the hotel had, but not all were scraped, or some other calculation.
- Total_Number_of_Reviews is not used in the model because of this unclarity.

In [7]:
# While there is an `Average_Score` for each hotel according to the dataset, 
# you can also calculate an average score (getting the average of all reviewer scores in the dataset for each hotel)
# Add a new column to your dataframe with the column header `Calc_Average_Score` that contains that calculated average. 

# define a function that takes a row and performs some calculation with it
def get_difference_review_avg(row):
  return row["Average_Score"] - row["Calc_Average_Score"]

# 'mean' is mathematical word for 'average'
df['Calc_Average_Score'] = round(df.groupby('Hotel_Name').Reviewer_Score.transform('mean'), 1)

# Add a new column with the difference between the two average scores
df["Average_Score_Difference"] = df.apply(get_difference_review_avg, axis = 1)

# Create a df without all the duplicates of Hotel_Name (so only 1 row per hotel)
review_scores_df = df.drop_duplicates(subset = ["Hotel_Name"])

# Sort the dataframe to find the lowest and highest average score difference
review_scores_df = review_scores_df.sort_values(by=["Average_Score_Difference"])

display(review_scores_df[["Average_Score_Difference", "Average_Score", "Calc_Average_Score", "Hotel_Name"]])

Average_Score_Difference  Average_Score  Calc_Average_Score  \
495945                      -0.8            7.7                 8.5   
111027                      -0.7            8.8                 9.5   
43688                       -0.7            7.5                 8.2   
178253                      -0.7            7.9                 8.6   
218258                      -0.5            7.0                 7.5   
...                          ...            ...                 ...   
151416                       0.7            7.8                 7.1   
22189                        0.8            7.1                 6.3   
250308                       0.9            8.6                 7.7   
68936                        0.9            6.8                 5.9   
3813                         1.3            7.2                 5.9   

                                               Hotel_Name  
495945                         Best Western Hotel Astoria  
111027  Hotel Stendhal Place Vend me Paris MGallery by...  
43688                       Mercure Paris Porte d Orleans  
178253                    Renaissance Paris Vendome Hotel  
218258                                Hotel Royal Elys es  
...                                                   ...  
151416                        Best Western Allegro Nation  
22189              Holiday Inn Paris Montparnasse Pasteur  
250308          MARQUIS Faubourg St Honor Relais Ch teaux  
68936                                       Villa Eugenie  
3813                                   Kube Hotel Ice Bar  

[1492 rows x 4 columns]

- You may also wonder about the Average_Score value and why it is sometimes different from the calculated average score.
- As we can't know why some of the values match, but others have a difference, it's safest in this case to use the review scores that we have to calculate the average ourselves.
- That said, the differences are usually very small.

- With only 1 hotel having a difference of score greater than 1, it means we can probably ignore the difference and use the calculated average score.

- Calculate and print out how many rows have column Negative_Review values of "No Negative"
- Calculate and print out how many rows have column Positive_Review values of "No Positive"
- Calculate and print out how many rows have column Positive_Review values of "No Positive" and Negative_Review values of "No Negative"

In [8]:
# with lambdas:
start = time.time()
no_negative_reviews = df.apply(lambda x: True if x['Negative_Review'] == "No Negative" else False , axis=1)
print("Number of No Negative reviews: " + str(len(no_negative_reviews[no_negative_reviews == True].index)))

no_positive_reviews = df.apply(lambda x: True if x['Positive_Review'] == "No Positive" else False , axis=1)
print("Number of No Positive reviews: " + str(len(no_positive_reviews[no_positive_reviews == True].index)))

both_no_reviews = df.apply(lambda x: True if x['Negative_Review'] == "No Negative" and x['Positive_Review'] == "No Positive" else False , axis=1)
print("Number of both No Negative and No Positive reviews: " + str(len(both_no_reviews[both_no_reviews == True].index)))
end = time.time()
print("Lambdas took " + str(round(end - start, 2)) + " seconds")

Number of No Negative reviews: 127890
Number of No Positive reviews: 35946
Number of both No Negative and No Positive reviews: 127
Lambdas took 10.6 seconds


Another way count items without Lambdas, and use sum to count the row

In [9]:
# without lambdas (using a mixture of notations to show you can use both)
start = time.time()
no_negative_reviews = sum(df.Negative_Review == "No Negative")
print("Number of No Negative reviews: " + str(no_negative_reviews))

no_positive_reviews = sum(df["Positive_Review"] == "No Positive")
print("Number of No Positive reviews: " + str(no_positive_reviews))

both_no_reviews = sum((df.Negative_Review == "No Negative") & (df.Positive_Review == "No Positive"))
print("Number of both No Negative and No Positive reviews: " + str(both_no_reviews))

end = time.time()
print("Sum took " + str(round(end - start, 2)) + " seconds")

Number of No Negative reviews: 127890
Number of No Positive reviews: 35946
Number of both No Negative and No Positive reviews: 127
Sum took 0.29 seconds


- You may have noticed that there are 127 rows that have both "No Negative" and "No Positive" values for the columns Negative_Review and Positive_Review respectively
- That means that the reviewer gave the hotel a numerical score, but declined to write either a positive or negative review.
- Luckily this is a small amount of rows (127 out of 515738, or 0.02%), so it probably won't skew our model or results in any particular direction, but you might not have expected a data set of reviews to have rows with no reviews, so it's worth exploring the data to discover rows like this.